### PR TITLE
Tdl 22157 support interrupted sync and bookmarking logic fix

### PR DESCRIPTION
**Pull request by rdeshmukh15**
# Description of change
Below are the changes done  [API doc link](https://api.iterable.com/api/docs):
1. Added interrupted sync support ([TDL-22157](https://jira.talendforge.org/browse/TDL-22157))
2. Fixed the bug where all records were not getting fetched in emails related streams ([TDL-22208](https://jira.talendforge.org/browse/TDL-22208))
3. Fixed `templates` and `campaigns` streams behaviour to make those work as INCREMENTAL streams ([TDL-22206](https://jira.talendforge.org/browse/TDL-22206))
4. Made the bookmark dates format uniform across all the streams ([TDL-22210](https://jira.talendforge.org/browse/TDL-22210))
5. Bookmarking logic bugfix ([TDL-22209](https://jira.talendforge.org/browse/TDL-22209), [TDL-22211](https://jira.talendforge.org/browse/TDL-22211))
6. Fix to parse the response of `list_users` stream properly ([TDL-22213](https://jira.talendforge.org/browse/TDL-22213))
7. Added unit test cases
8. Changes in config.yml

# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_iterable/helper.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+def epoch_to_datetime_string(milliseconds):
+    """Function to convert epoch time to datetime """
+    datetime_string = None
+    try:
+        datetime_string = time.strftime("%Y-%m-%d %H:%M:%S %Z", time.localtime(milliseconds / 1000))
+    except TypeError:
+        # If fails, it means format already datetime string.
+        datetime_string = milliseconds
+    return datetime_string
```
Comment (sgandhi1311) [inline tap_iterable/helper.py:12]: Newline at end of file.

File: tap_iterable/__init__.py [hunk 2]
```diff
@@ -26,50 +25,13 @@
...
-        counter_value = sync_stream(state, instance)
-        logger.info("%s: Completed sync (%s rows)", stream_name, counter_value)
-
-    singer.write_state(state)
-    logger.info("Finished sync")
-
-
-@singer.utils.handle_top_exception(logger)
+@singer.utils.handle_top_exception(LOGGER)
 def main():
```
Comment (sgandhi1311) [inline tap_iterable/__init__.py:35]:
```suggestion
@singer.utils.handle_top_exception(logger)
def main():
```
Do u think to have this decorator back in place?

**Review by sgandhi1311 — COMMENTED** [review]
Left a review comment

File: tap_iterable/iterable.py [hunk 3]
```diff
@@ -137,11 +142,19 @@ def templates(self, column_name, bookmark):
       "InApp",
       "SMS"
     ]
+    # bookmark value `updatedAt` is getting round off to next second
+    # while putting into the STATE and which is causing 0 records fetched
+    # with existing bookmark value in STATE 
+    bookmark_val = utils.strptime_with_tz(bookmark)- timedelta(seconds=1)
```
Comment (leslievandemark) [inline tap_iterable/iterable.py:148]: is this only a bug for templates stream?

File: tap_iterable/__init__.py (hunk 2 above)
Comment (sgandhi1311) [reply tap_iterable/__init__.py:35]: Please ignore the above comment. I see the decorator is already in place, couldn't see it in github.dev

File: tap_iterable/iterable.py (hunk 3 above)
Comment (rdeshmukh15) [reply tap_iterable/iterable.py]: Yes, this is the bug with `templates` stream only.
Actually, I have added the millisecond level granularity to maintain the bookmark value.
So, before I was maintaining bookmark date like this "updatedAt": "2023-02-22T07:31:15.000000Z"
And, now, I'm storing it like "updatedAt": "2023-02-22T07:31:15.145000Z"
Still, the issue persists and it is actually API side bug that if we keep `startDateTime`= "2023-03-01+07%3A31%3A15"
It doesn't fetch the records with updatedAt= "2023-03-01 07:31:15" though record with that time present for the stream template. And, hence, subtracted 1 second from the existing bookmark value and made it to updatedAt= "2023-03-01 07:31:14" to fetch at least 1 record.

**Review by sgandhi1311 — APPROVED** [review]

**Review by leslievandemark — APPROVED** [review]

**Pull request merged**